lib/graphql/schema/build_from_definition.rb

```diff
@@ -111,11 +115,11 @@ def build(document, default_resolve:, using: {}, interpreter: true, relay:)
             end
 
             if default_resolve.respond_to?(:resolve_type)
-              define_singleton_method(:resolve_type) do |*args|
-                default_resolve.resolve_type(*args)
+              def self.resolve_type(*args)
+                self.definition_default_resolve.resolve_type(*args)
               end
             else
-              define_singleton_method(:resolve_type) do |*args|
+              def self.resolve_type(*args)
```

Comment:
By removing all the retained procs, lots of the intermediate objects of this process were GC'ed instead of retained (by the proc binding).